File: Data Structures & Algorithms/binary-tree-diameter/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:

    def maxHeight(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0
        leftHeight = self.maxHeight(root.left)
        rightHeight = self.maxHeight(root.right)
        self.dia = max(self.dia, leftHeight + rightHeight)
        return max(leftHeight, rightHeight) + 1

    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0
        self.dia = 0
        self.maxHeight(root)
        return self.dia

File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int dfs(TreeNode* root, int& maxSum) {
        if(!root) return 0;
        int lSum = max(0, dfs(root -> left, maxSum));
        int rSum = max(0, dfs(root -> right, maxSum));
        maxSum = max(maxSum, lSum + rSum + root -> val);
        return root -> val + max(lSum, rSum);
    }

    int maxPathSum(TreeNode* root) {
        int maxSum = root -> val;
        dfs(root, maxSum);
        return maxSum;
    }
};
